python/integration_tests/test_upkeep_retry.py

```diff
@@ -0,0 +1,337 @@
+            else:
+                next_task = worker.complete_task(task)
+    except Exception as e:
+        print(f"[taskworker_{worker_id}]: Worker process crashed: {e}")
+        return
+
+    with open(log_file_path, "a") as log_file:
+        log_file.write(f"Retried:{retried_tasks}")
+
+
+def manage_consumer(
+    consumer_path: str,
+    config_file_path: str,
+    consumer_config: dict,
+    log_file_path: str,
+    timeout: int,
+    num_messages: int,
+    tasks_written_event: threading.Event,
+    shutdown_events: list[threading.Event],
+) -> None:
+    with open(log_file_path, "a") as log_file:
+        print(f"[consumer_0] Starting consumer, writing log file to {log_file_path}")
+        process = subprocess.Popen(
+            [consumer_path, "-c", config_file_path],
+            stderr=subprocess.STDOUT,
+            stdout=log_file,
+        )
+        time.sleep(3)  # give the consumer some time to start
+
+        # Let the consumer write the messages to sqlite
+        end = time.time() + timeout
+        while (time.time() < end) and (not tasks_written_event.is_set()):
+            written_tasks = check_num_tasks_written(consumer_config)
+            if written_tasks == num_messages:
+                print(
+                    f"[consumer_0]: Finishing writting all {num_messages} task(s) to sqlite. Sending signal to taskworker(s) to start processing"
+                )
+                tasks_written_event.set()
+            time.sleep(1)
+
+        # Keep gRPC consumer alive until taskworker is done processing
+        if tasks_written_event.is_set():
+            print("[consumer_0]: Waiting for taskworker(s) to finish processing...")
+            while not all(
+                shutdown_event.is_set() for shutdown_event in shutdown_events
+            ):
+                time.sleep(1)
+            print("[consumer_0]: Received shutdown signal from all taskworker(s)")
+        else:
+            print(
+                "[consumer_0]: Timeout elapse and not all tasks have been written to sqlite. Signalling taskworker(s) to stop"
+            )
+            for shutdown_event in shutdown_events:
+                shutdown_event.set()
+
+        # Stop the consumer
+        print("[consumer_0]: Shutting down consumer")
+        process.send_signal(signal.SIGINT)
+        try:
+            return_code = process.wait(timeout=10)
+            assert return_code == 0
+        except Exception:
+            process.kill()
+
+
+def check_num_tasks_written(consumer_config: dict) -> int:
+    attach_db_stmt = f"ATTACH DATABASE '{consumer_config['db_path']}' AS {consumer_config['db_name']};\n"
+    query = f"""SELECT count(*) as count FROM {consumer_config['db_name']}.inflight_taskactivations;"""
+    con = sqlite3.connect(consumer_config["db_path"])
+    cur = con.cursor()
+    cur.executescript(attach_db_stmt)
+    rows = cur.execute(query).fetchall()
+    count = rows[0][0]
+    return count
```

Comment:
This function shows up in a few integration tests, perhaps we should have a module of common integration test helpers?

Reply:
Good catch, lifted the function and consumer configs to helpers.py